ql/math/array.cpp

```diff
@@ -0,0 +1,618 @@
+        detail::_fill_array_(*this, data_, n_, init.begin(), init.end(),
+                             std::false_type());
+    }
+
+    Array& Array::operator=(const Array& from) {
+        // strong guarantee
+        Array temp(from);
+        swap(temp);
+        return *this;
+    }
+
+    Array& Array::operator=(Array&& from) noexcept {
+        swap(from);
+        return *this;
+    }
+
+    bool Array::operator==(const Array& to) const {
+        return (n_ == to.n_) && std::equal(begin(), end(), to.begin());
+    }
+
+    bool Array::operator!=(const Array& to) const {
+        return !(this->operator==(to));
+    }
+
+    const Array& Array::operator+=(const Array& v) {
+        QL_REQUIRE(n_ == v.n_,
+                   "arrays with different sizes (" << n_ << ", "
+                   << v.n_ << ") cannot be added");
+        std::transform(begin(),end(),v.begin(),begin(), std::plus<>());
+        return *this;
+    }
+
+
+    const Array& Array::operator+=(Real x) {
+        std::transform(begin(), end(), begin(), [=](Real y) -> Real { return y + x; });
+        return *this;
+    }
+
+    const Array& Array::operator-=(const Array& v) {
+        QL_REQUIRE(n_ == v.n_,
+                   "arrays with different sizes (" << n_ << ", "
+                   << v.n_ << ") cannot be subtracted");
+        std::transform(begin(), end(), v.begin(), begin(), std::minus<>());
+        return *this;
+    }
+
+    const Array& Array::operator-=(Real x) {
+        std::transform(begin(),end(),begin(), [=](Real y) -> Real { return y - x; });
+        return *this;
+    }
+
+    const Array& Array::operator*=(const Array& v) {
+        QL_REQUIRE(n_ == v.n_,
+                   "arrays with different sizes (" << n_ << ", "
+                   << v.n_ << ") cannot be multiplied");
+        std::transform(begin(), end(), v.begin(), begin(), std::multiplies<>());
+        return *this;
+    }
+
+    const Array& Array::operator*=(Real x) {
+        std::transform(begin(), end(), begin(), [=](Real y) -> Real { return y * x; });
+        return *this;
+    }
+
+    const Array& Array::operator/=(const Array& v) {
+        QL_REQUIRE(n_ == v.n_,
+                   "arrays with different sizes (" << n_ << ", "
+                   << v.n_ << ") cannot be divided");
+        std::transform(begin(), end(), v.begin(), begin(), std::divides<>());
+        return *this;
+    }
+
+    const Array& Array::operator/=(Real x) {
+        #if defined(QL_EXTRA_SAFETY_CHECKS)
+        QL_REQUIRE(x != 0.0, "division by zero");
+        #endif
+        std::transform(begin(), end(), begin(), [=](Real y) -> Real { return y / x; });
+        return *this;
+    }
+
+    Real Array::operator[](Size i) const {
+        #if defined(QL_EXTRA_SAFETY_CHECKS)
+        QL_REQUIRE(i<n_,
+                   "index (" << i << ") must be less than " << n_ <<
+                   ": array access out of range");
+        #endif
+        return data_.get()[i];
+    }
```

Comment:
Let's at least inline element access, please.  I'd also try timing some heavy finite-difference calculations to check that we're not pessimizing run time.